pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -9,89 +9,189 @@ import (
 	"github.com/prometheus/prometheus/promql"
 
 	"github.com/grafana/mimir/pkg/streamingpromql/types"
-	"github.com/grafana/mimir/pkg/util/limiter"
 )
 
+type undoAction int
+
+const (
+	none undoAction = iota
+	remove
+	replace
+)
+
+// AnchoredExtensionMetadata tracks the first and last points considered in a range extension.
+// It also tracks modifications made to the buffer, such as synthetic boundary points and head/tail trimming.
+// A function is provided to undo these changes on a given buffer.
+type AnchoredExtensionMetadata struct {
+	// The original first point in the set of points considered for the anchored extension. This may not be the first point in the overall buffer.
+	first promql.FPoint
+	// The original last point in the set of points considered for the anchored extension. This may not be the last point in the overall buffer.
+	last promql.FPoint
+	// If the buffer has a trailing point which is outside consideration for the anchored extension it is stored here so it can be restored.
+	excludedLast promql.FPoint
+	// Flag indicating that the point in excludedLast needs to be re-inserted to the end of the buffer.
+	restoreExcludedLast bool
+	// If the buffer has a leading point which is outside consideration for the anchored extension it is stored here so it can be restored.
+	// This can occur when there is a single point in the extended look-back. The point can not be used since there is no point within the original time range.
+	// But we can not delete it from the buffer since the point may be used in subsequent step iterations.
+	excludedFirst        promql.FPoint
+	restoreExcludedFirst bool
+	// The actions required to undo synthetic head and tail modifications.
+	undoTailModifications undoAction
+	undoHeadModifications undoAction
+}
+
+// UndoChanges will restore the buffer to its original points.
+func (m *AnchoredExtensionMetadata) UndoChanges(buff *types.FPointRingBuffer) error {
```

Comment:
Should we have a protection when UndoChanges is applied multiple times?  What is the expectation when such operation happens?